internal/lib/stats/stats_server_linux.go

```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
 	})
 }
 
+// metricsForPodSandbox is an internal, non-locking version of MetricsForPodSandbox
+// that returns (and occasionally gathers) the metrics for the given sandbox.
+// Note: caller must hold the lock on the StatsServer
+func (ss *StatsServer) metricsForPodSandbox(sb *sandbox.Sandbox) *SandboxMetrics {
+	if ss.collectionPeriod == 0 {
+		return ss.updatePodSandboxMetrics(sb)
+	}
+	sboxMetrics, ok := ss.sboxMetrics[sb.ID()]
+	if ok {
+		return sboxMetrics
+	}
+	// Cache miss, try again
+	return ss.updatePodSandboxMetrics(sb)
+}
+
+// updatePodSandboxMetrics updates the sandbox metrics for the given sandbox.
+// If the sandbox is not found, it creates a new entry in the map.
+// Note: caller must hold the lock on the StatsServer.
+func (ss *StatsServer) updatePodSandboxMetrics(sb *sandbox.Sandbox) *SandboxMetrics {
+	if sb == nil {
+		return nil
+	}
+	sm, exists := ss.sboxMetrics[sb.ID()]
+	if !exists {
+		sm = NewSandboxMetrics(sb)
+	}
+	// network metrics are collected at pod level
+	if slices.Contains(ss.Config().IncludedPodMetrics, "network") {
+		podMetrics := ss.GenerateNetworkMetrics(sb)
+		sm.metric.Metrics = podMetrics
+	}
+
+	containersList := sb.Containers().List()
+	containerMetrics := make([]*types.ContainerMetrics, 0, len(containersList))
+	for _, c := range containersList {
+		// Skip if the container is stopped
+		if c.StateNoLock().Status == oci.ContainerStateStopped {
+			continue
+		}
+		cMetrics := ss.GenerateSandboxContainerMetrics(sb, c, sm)
+		containerMetrics = append(containerMetrics, cMetrics)
+	}
+	sm.metric.ContainerMetrics = containerMetrics
+	ss.sboxMetrics[sb.ID()] = sm // Update the entry in the map
```

Comment:
No need to explain obvious code.

```suggestion
	ss.sboxMetrics[sb.ID()] = sm
```